### PR TITLE
Rework font loading to use file paths instead of font names

### LINKED ISSUE #40774: Android bug and Hungarian language
and i want to report 2 android bugs.
1. The software rendering is faulty so far but now it throws out some device I am trying to boot from.
Screenshot_20200523-163752

2. After loading the second game, the Hungarian language is displayed incorrectly and the Hungarian characters are replaced with characters, thus making the whole thing unreadable
Screenshot_20200523-163421
. Thank you for reading it, I hope to correct it as soon as possible.

### LINKED ISSUE #42638: Some non-Ascii characters are not shown correctly
# Describe the bug

Non-Ascii characters, like the german umlauts (ä, ö, ü) and some characters from other languages like the ñ as in español are shown either as some other characters or just as an empty box. (See screenshots below)

I suspect this is probably an issue with the encoding.

I have tested this on Windows, Android and Linux, but I can only reproduce it on Windows 10 and Android. Linux seems not to be affected by this issue.

# Steps To Reproduce

**Steps to reproduce for the screenshot below:**
1. Go to your settings under Interface -> Language and set it to german (Deutsch).
2. Open up the settings menu again.
3. There will be some texts that are not displayed correctly.

**Alternative steps to reproduce:**
1. Go to your settings under Interface -> Language.
2. You will see that in the description at the top where the languages are listed, some words like español will not be displayed correctly.

# Expected behavior

Those characters should be displayed correctly.

# Screenshots

I have underlined the incorrectly displayed characters in red.

optionsMenu

ingame

small

On android the characters look the same:
android

# Versions and configuration

 - OS: Windows 10, Android 10 and also tested under openSuse 15.1, but the bug does not happen there.
 - Game Version: latest (0.E-4667-g1d4ea08)
 - Graphics version: Tiles
 - Ingame language: German

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
#### Summary

```SUMMARY: Infrastructure "Rework font loading to use file paths instead of font names"```

#### Purpose of change

Fixes #42638
Fixes #40774

#### Describe the solution

Added TTF version of Terminus font.

Fonts are now stored in the config and loaded using file paths instead of storing font names and searching for corresponding file  in `fontlist.txt`.

Legacy config compatibility is preserved for non-system fonts - e.g. if legacy config file contains font name `Terminus`, game would try to find font file with that name (using file extensions `ttf` and then `fon`) in following folders:

- `data/font`;
- `./font`;
- `%systemroot%\fonts` (if you are on Windows);
- `/usr/share/fonts` (if you are on Linux);
- `/usr/local/share/fonts` (if you are on Linux);
- `/.fonts` (if you are on Linux and `HOME` env var is set);
- `/System/Library/Fonts` (if you are on MacOS);
- `/Library/Fonts` (if you are on MacOS);
- `~/Library/Fonts` (if you are on MacOS).

And fall back to `data/font/unifont.ttf` if nothing else is found.

#### Describe alternatives you've considered

Only change font file (#42827).

#### Testing

Acute characters and umlauts are displayed correctly with default or legacy config.

#### Additional context

- compared to alternative, I've preserved `terminus.fon` as it works fine for English language.

### CODE REVIEW

File: src/font_loader.h [hunk 1]
```diff
@@ -14,6 +14,14 @@
 #include "path_info.h"
 #include "cata_utility.h"
 
+// Ensure that unifont is always loaded as a fallback font to prevent users from shooting themselves in the foot
+static auto ensure_unifont_loaded = []( std::vector<std::string> &font_list )
```
Comment (Qrox) [inline src/font_loader.h:18]: This can be a normal function instead of a lambda.

File: src/font_loader.h [hunk 2]
```diff
@@ -14,6 +14,14 @@
...
+{
+    if( std::find( std::begin( font_list ), std::end( font_list ), "unifont" ) == font_list.end() ) {
+        font_list.emplace_back( PATH_INFO::fontdir() + "unifont.ttf" );
+    }
+};
```
Comment (ZhilkinSerg) [inline src/font_loader.h:23]:
```suggestion
static void ensure_unifont_loaded( std::vector<std::string> &font_list )
{
    if( std::find( std::begin( font_list ), std::end( font_list ), "unifont" ) == font_list.end() ) {
        font_list.emplace_back( PATH_INFO::fontdir() + "unifont.ttf" );
    }
}
```

File: src/font_loader.h (hunk 1 above)
Comment (ZhilkinSerg) [reply src/font_loader.h]: Thanks I forgot to change it.

File: src/sdl_font.cpp [hunk 3]
```diff
@@ -323,30 +187,82 @@ CachedTTFFont::CachedTTFFont(
...
+    wordexp_t exp;
+    wordexp( "~/Library/Fonts/", &exp, 0 );
+    known_prefixes.emplace_back( exp.we_wordv[0] );
+    wordfree( &exp );
+    */
+#else // Other POSIX-ish systems
+    known_prefixes.emplace_back( "/usr/share/fonts/" );
+    known_prefixes.emplace_back( "/usr/local/share/fonts/" );
+    char *home;
+    if( home = getenv( "HOME" ) ) {
```
Comment (ZhilkinSerg) [inline src/sdl_font.cpp:229]:
```suggestion
    if( ( home = getenv( "HOME" ) ) ) {
```

File: src/sdl_font.cpp [hunk 4]
```diff
@@ -323,30 +187,82 @@ CachedTTFFont::CachedTTFFont(
...
+        typefaces.emplace_back( typeface );
+    }
+    ensure_unifont_loaded( typefaces );
+
+    for( const std::string &tf : typefaces ) {
+        if( !file_exist( tf ) ) {
+            dbg( D_ERROR ) << "Not found " << tf;
+            continue;
+        }
+        dbg( D_ERROR ) << "Loading truetype font " << tf;
```
Comment (olanti-p) [inline src/sdl_font.cpp:261]: `D_ERROR` makes the game emit a backtrace here. Was it supposed to be `D_INFO` or something?

Comment (ZhilkinSerg) [reply src/sdl_font.cpp:261]: > 
> 
> `D_ERROR` makes the game emit a backtrace here. Was it supposed to be `D_INFO` or something?

Yes, log level should've been reduced.

**Pull request merged**